File: 3003-maximize-the-number-of-partitions-after-operations/3003-maximize-the-number-of-partitions-after-operations.py
class Solution:
    def maxPartitionsAfterOperations(self, s: str, k: int) -> int:
        n = len(s)
        @lru_cache(None)
        def dfs(idx,mask,canChange):
            # return the max partitions from idx to end in string s 
            if idx == n: return 0
            
            res = 0
            curCharIdx = ord(s[idx])-ord('a')
            newMask = mask | (1<<curCharIdx)
            distinctCount = newMask.bit_count()
            if distinctCount>k:
                res = 1 + dfs(idx+1, 1<<curCharIdx, canChange)
            else:
                res = dfs(idx+1, newMask, canChange)
            if canChange:
                # try change curChar to any of the 26
                for i in range(26):
                    if i==curCharIdx:continue
                    newMask = mask | (1<<i)
                    distinctCount = newMask.bit_count()
                    if distinctCount>k:
                        res = max(res, 1 + dfs(idx+1,1<<i,False))
                    else:
                        res = max(res, dfs(idx+1,newMask,False))
            return res
        return dfs(0,0,True)+1
                
            
                
            
            

File: 3003-maximize-the-number-of-partitions-after-operations/3003-maximize-the-number-of-partitions-after-operations.java
class Solution {
    int k;
    String s;    
    HashMap<Long,Integer> dp;
    public int maxPartitionsAfterOperations(String s, int k) {
        this.k = k;
        this.s = s;
        this.dp = new HashMap<>();
        return dfs(0, 0, true)+1;
    }
    private int dfs(int idx, int mask,boolean canchange) {
        // return: max number of partitions s[0:idx]
        
        long key = ((long)idx<<27|((long)mask<<1)|(canchange?1:0));
        if (dp.containsKey(key))            
            return dp.get(key);
        if (idx == s.length())
            return 0;
        int newmask = mask | (1<<(s.charAt(idx)-'a'));
        int ret = 0;
        if (Integer.bitCount(newmask)>k) {
            ret = 1 + dfs(idx + 1, (1<<(s.charAt(idx)-'a')),canchange);
        }
        else {
            ret = dfs(idx + 1, newmask,canchange);
        }

        if (canchange) {
            for (int i = 0; i<26; i++) {
                newmask = mask | (1<<i);
                if (Integer.bitCount(newmask)>k) {
                    ret = Math.max(ret, 1 + dfs(idx + 1, (1<<i), false));
                }
                else {
                    ret = Math.max(ret, dfs(idx + 1, newmask,false));
                }
            }
        }
        dp.put(key, ret);
        return ret;
    }
}